### PR TITLE
Turning hal.interface into attributes.

### PR DESCRIPTION
**Pull request by benvanik**
This avoids the need for hal.interface lookups as the places that
require the layout information now embed it directly. This avoids
the need for symbol table walking and other rough edges.

This only works because we perform all layout optimization prior to
materializing the attributes (which would be really difficult to update
in the whole program in this form). Thanks `stream` dialect.

This did expose a lot of codegen tests that had incorrect interfaces -
I've tried to fix most of them up here but will wait until this settles to
enable verification so they can't drift again.

### CODE REVIEW

File: iree/compiler/Codegen/LLVMCPU/ConvertToLLVM.cpp [hunk 1]
```diff
@@ -750,16 +736,12 @@ void ConvertToLLVMPass::runOnOperation() {
...
+    ConvertHALInterfaceBindingSubspanOp
   >(&getContext(), converter);
   // clang-format on
 
   LLVMConversionTarget target(getContext());
-  // IREE::HAL::InterfaceOp will be removed after successful conversion of the
-  // rest of the IR.
-  target.addLegalOp<ModuleOp, IREE::HAL::InterfaceOp,
-                    IREE::HAL::InterfaceBindingOp, IREE::HAL::InterfaceEndOp>();
+  target.addLegalOp<ModuleOp>();
```
Comment (ScottTodd) [inline iree/compiler/Codegen/LLVMCPU/ConvertToLLVM.cpp:744]: nice cleanup :)

File: iree/compiler/ConstEval/Runtime.cpp [hunk 2]
```diff
@@ -102,7 +102,7 @@ LogicalResult CompiledBinary::invokeNullary(Location loc, StringRef name,
   if (auto status = iree_vm_module_lookup_function_by_name(
           main_module, IREE_VM_FUNCTION_LINKAGE_EXPORT,
           iree_string_view_t{name.data(), name.size()}, &function)) {
-    iree_status_consume_code(status);
+    iree_status_ignore(status);
```
Comment (ScottTodd) [inline iree/compiler/ConstEval/Runtime.cpp:105]: unrelated to the rest of this PR, but fine..

File: iree/task/worker.c [hunk 3]
```diff
@@ -263,10 +263,10 @@ static bool iree_task_worker_pump_once(
   // Since we can host work from multiple scopes and want to ensure an error
   // in one doesn't bring down the whole system we pretend we executed
   // something here by falling through.
-  IREE_ASSERT_TRUE(iree_status_is_ok(status));
   if (!iree_status_is_ok(status)) {
     iree_status_fprint(stderr, status);
   }
+  IREE_ASSERT_TRUE(iree_status_is_ok(status));
```
Comment (ScottTodd) [inline iree/task/worker.c:269]: also unrelated, but fine :P

**Review by ScottTodd — APPROVED** [review]
lgtm after build/tests pass

File: iree/task/worker.c (hunk 3 above)
Comment (benvanik) [reply iree/task/worker.c:269]: I ended up needing to debug through this and was finding that having the actual status printed _before_ the debugger took over was useful :P

**Pull request merged**